packages/storage/src/adapters/S3Storage/S3Storage.ts

```diff
@@ -0,0 +1,181 @@
+  private partSize?: number
+  /**
+   * The tags to apply to the object.
+   */
+  private tags: Tag[]
+
+  constructor(opts: S3StorageOptions) {
+    super(opts)
+
+    const { queueSize, partSize, tags } = opts
+
+    this.s3Client = new S3Client({
+      region: opts.region,
+      credentials: {
+        accessKeyId: opts.accessKeyId,
+        secretAccessKey: opts.secretAccessKey,
+      },
+      endpoint: opts.endpoint,
+      forcePathStyle: !!opts.endpoint,
+    })
+    this.bucket = opts.bucket
+    this.cacheControl = opts.cacheControl
+    this.tags = tags || []
+    this.queueSize = queueSize
+    this.partSize = partSize
+    this.tags = opts.tags || []
+    this.showProgress = opts.showProgress || false
+  }
+
+  /**
+   * Saves a file to the S3 storage.
+   *
+   * The key is the path to the file in the S3 bucket and is constructed from
+   * the baseDir, the file name, and the path override.
+   *
+   * Use the `saveOverride` parameter to specify a path override to organize
+   * your files in a specific folder in the bucket.
+   *
+   * Default params:
+   * - ContentType: The MIME type of the file.
+   * - CacheControl: 'max-age=604800' // 1 week
+   *
+   * @param file - The file to save.
+   * @param saveOverride - Optional overrides for the save operation.
+   * @returns The location of the saved file.
+   */
+  async save(file: File, saveOverride?: SaveOptionsOverride) {
+    const fileName = this.generateFileNameWithExtension(saveOverride, file)
+    const key = `${saveOverride?.path || this.adapterOpts.baseDir}/${fileName}`
+
+    const upload = new Upload({
+      params: {
+        ContentType: file.type,
+        CacheControl: this.cacheControl || 'max-age=604800',
+        Bucket: this.bucket,
+        Key: key,
+        Body: file,
+      },
+      client: this.s3Client,
+      queueSize: this.queueSize,
+      partSize: this.partSize,
+      tags: this.tags,
+    })
+
+    if (this.showProgress) {
+      upload.on('httpUploadProgress', (progress) => {
+        console.log(progress)
+      })
+    }
+
+    const result = await upload.done()
+    // The result looks like:
+    //    {
+    //  '$metadata': {
+    //    httpStatusCode: 200,
+    //    requestId: '992625086802962799',
+    //      extendedRequestId: undefined,
+    //        cfId: undefined,
+    //      attempts: 1,
+    //      totalRetryDelay: 0
+    //   },
+    //   ETag: '"a17fabeb1fcf56b1022a8d236374066e"',
+    //   Bucket: 'bucket',
+    //   Key: 'baseDir/path/fileName.png',
+    //   Location: 'https://fly.storage.tigris.dev/rw-showcases/baseDir/path/fileName.png'
+    // }
+    // We save key because that what the S3UrlSigner and the S3 client expects when signing or reading
+    // The result location is the url in S3 but for private buckets you cannot access that without the signed url
+    // So, therefore we save the key and return that as the location
+    // TODO: Should we check the status code and throw an error if it's not 200?
```

Comment:
I think it would be good to figure out the answer to this TODO before we merge this PR

Reply:
Checking status now